source/iterator/snapshot.go

```diff
@@ -0,0 +1,234 @@
+func (s *snapshot) loadRecords(ctx context.Context) error {
+	selectDataset := goqu.Dialect("mysql").Select()
+
+	if len(s.columns) > 0 {
+		keyColumnPresent := false
+		// add one to the capacity to have a space for the keyColumn
+		// if it's not present in the columns list.
+		cols := make([]any, 0, len(s.columns)+1)
+		for i := 0; i < len(s.columns); i++ {
+			if s.columns[i] == s.keyColumn {
+				keyColumnPresent = true
+			}
+
+			cols = append(cols, s.columns[i])
+		}
+
+		if !keyColumnPresent {
+			cols = append(cols, s.keyColumn)
+		}
+
+		selectDataset = selectDataset.Select(cols...)
+	}
+
+	selectDataset = selectDataset.
+		From(s.table).
+		Order(goqu.C(s.orderingColumn).Asc()).
+		Limit(uint(s.batchSize))
+
+	if s.position != nil {
+		selectDataset = selectDataset.Where(
+			goqu.Ex{
+				s.orderingColumn: goqu.Op{"gt": s.position.LastProcessedElementValue},
+			},
+		)
+	}
+
+	query, _, err := selectDataset.ToSQL()
+	if err != nil {
+		return fmt.Errorf("construct select query: %w", err)
+	}
+
+	resultStream, err := s.session.StreamExecute(ctx, query, nil)
+	if err != nil {
+		return fmt.Errorf("stream execute select query: %w", err)
+	}
+
+	if err := s.processStreamResults(ctx, resultStream); err != nil {
+		return fmt.Errorf("process stream results: %w", err)
+	}
+
+	return nil
+}
+
+// processStreamResults receives a result from the sqltypes.ResultStream,
+// converts its rows to the sdk.Record and sends them to the i.records channel.
+func (s *snapshot) processStreamResults(ctx context.Context, resultStream sqltypes.ResultStream) error {
+	for {
+		result, err := resultStream.Recv()
+		if err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
+			}
+
+			return fmt.Errorf("stream recv: %w", err)
+		}
+
+		if result.Fields != nil {
+			s.fields = result.Fields
+		}
+
+		for _, row := range result.Rows {
+			if s.fields == nil {
+				// shouldn't happen cause VEventType_FIELD always comes before VEventType_ROW.
+				sdk.Logger(ctx).Warn().Msgf("i.fields is nil, skipping the row")
+
+				continue
+			}
+
+			transformedRow, err := columntypes.TransformValuesToNative(ctx, s.fields, row)
+			if err != nil {
+				return fmt.Errorf("transform row: %w", err)
+			}
+
+			s.position = &Position{
+				Mode:                      ModeSnapshot,
+				Keyspace:                  s.keyspace,
+				LastProcessedElementValue: transformedRow[s.orderingColumn],
+			}
+
+			sdkPosition, err := s.position.MarshalSDKPosition()
+			if err != nil {
+				return fmt.Errorf("marshal position: %w", err)
+			}
+
+			transformedRowBytes, err := json.Marshal(transformedRow)
+			if err != nil {
+				return fmt.Errorf("marshal row: %w", err)
+			}
+
+			metadata := make(sdk.Metadata)
+			metadata.SetCreatedAt(time.Now())
+			metadata[metadataKeyTable] = s.table
+
+			s.records <- sdk.Util.Source.NewRecordSnapshot(sdkPosition, metadata, sdk.StructuredData{
+				s.keyColumn: transformedRow[s.keyColumn],
+			}, sdk.RawData(transformedRowBytes))
```

Comment:
Maybe returning sdk.StructuredData would be more natural?

Reply:
Updated.